odoo/addons/base/models/res_users.py

```diff
@@ -786,6 +786,61 @@ def change_password(self, old_passwd, new_passwd):
         # use self.env.user here, because it has uid=SUPERUSER_ID
         return self.env.user.write({'password': new_passwd})
 
+    @api.model
+    def _deactivate_portal_user(self):
+        """Try to remove the current portal user.
+
+        If it failed, archive the user instead and remove sensitive information.
+
+        This is used to give the opportunity to portal users to de-activate their accounts.
+        Indeed, as the portal users can easily create accounts, they sometimes will wish
+        it removed because they don't use this Odoo portal anymore.
+
+        Before this feature, they would have to contact the website or the support to get
+        their account removed, which could be tedious.
+        """
+        if not self.env.user.has_group('base.group_portal'):
+            raise UserError(_('Only the portal users can deactivate their accounts.'))
+
+        ip = request.httprequest.environ['REMOTE_ADDR'] if request else 'n/a'
+        _logger.info('Account deletion asked for "%s" (#%i) from %s', self.env.user.login, self.env.uid, ip)
+
+        log_user = f'User #{self.env.uid}:'
+
+        # A user can not self-deactivate, switch to super user
+        user = self.env.user.with_user(SUPERUSER_ID)
+
+        try:
+            self._cr.execute('SAVEPOINT remove_portal_user')
+            partner = user.partner_id
+            user.unlink()
+            partner.unlink()
```

Comment:
FTR, these deletions are prohibitively slow on a large database, because the DB needs to verify/nullify every single FK that exists in the database. For `user` this includes *all* the `create_uid`/`write_uid` FK in all tables. Of course those columns aren't generally indexed, so that can easily take 5-10 minutes.

Some possible strategies to mitigate this:
- A. add indexes on the biggest tables for create_uid, write_uid, user_id and partner_id
- B. avoid deleting users immediately, always archive them first, and create a "deletion queue" to be processed later through a cron
- C. or change the ORM to not create the FK constraints on the create_uid/write_uid columns, deciding that it's ok for those become orphan (but that's a big hairy change, which the ORM may need to hide to userland)

Reply:
@odony 
cc @awa-odoo 

Hum ok, I vote for the cron as it's the solution that will make less changes in others model, but it will cost one new field on res.users to know which record we have to delete in the cron (and when the deletion failed to not re-try it)

Reply:
I'd suggest creating a `res.users.deletion` table rather than adding fields on res_users. In addition to avoiding bloat on `res.users`, it would serve as a useful history log of the deletion requests.

- If they fail, we can keep track of the failure there and not retry it (or even manually re-schedule for deletion later, manually, if we change something).
- If they succeed, we could keep the anonymous info there, e.g. `user_id` (int, not m2o), `create_date`, `state` is ok to keep, even for GDPR. We should distinguish `user_id` and `create_uid`, as it could make sense for admins to be able to request deletion for their users, and it's nice to keep track of who did what. And `create_uid` will be NULL if the deletion succeeds, if the user requested it themselves.